mono/utils/flags.c

```diff
@@ -0,0 +1,217 @@
+/**
+ * \file Runtime flags
+ *
+ * Copyright 2020 Microsoft
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+#include <stdio.h>
+
+#include "flags.h"
+#include "mono/utils/mono-error-internals.h"
+
+typedef enum {
+	MONO_FLAG_BOOL,
+	MONO_FLAG_BOOL_READONLY,
+	MONO_FLAG_INT,
+	MONO_FLAG_STRING
+} MonoFlagType;
+
+/* Define flags */
+#define DEFINE_FLAG_FULL(flag_type, ctype, c_name, cmd_name, def_value, comment) \
+	ctype mono_##c_name = def_value;
+#define DEFINE_FLAG_READONLY(flag_type, ctype, c_name, cmd_name, def_value, comment)
+#include "flag-definitions.h"
+
+/* Flag metadata */
+typedef struct {
+	MonoFlagType flag_type;
+	gpointer addr;
+	const char *cmd_name;
+} FlagData;
+
+static FlagData flag_meta[] = {
+#define DEFINE_FLAG_FULL(flag_type, ctype, c_name, cmd_name, def_value, comment) \
+	{ flag_type, &mono_##c_name, cmd_name },
+#define DEFINE_FLAG_READONLY(flag_type, ctype, c_name, cmd_name, def_value, comment) \
+	{ flag_type, NULL, cmd_name },
+#include "flag-definitions.h"
+};
+
+static const char*
+flag_type_to_str (MonoFlagType type)
+{
+	switch (type) {
+	case MONO_FLAG_BOOL:
+		return "bool";
+	case MONO_FLAG_BOOL_READONLY:
+		return "bool (read-only)";
+	case MONO_FLAG_INT:
+		return "int";
+	case MONO_FLAG_STRING:
+		return "string";
+	default:
+		g_assert_not_reached ();
+		return NULL;
+	}
+}
+
+static char *
+flag_value_to_str (MonoFlagType type, gconstpointer addr)
+{
+	switch (type) {
+	case MONO_FLAG_BOOL:
+	case MONO_FLAG_BOOL_READONLY:
+		return *(gboolean*)addr ? g_strdup ("true") : g_strdup ("false");
+	case MONO_FLAG_INT:
+		return g_strdup_printf ("%d", *(int*)addr);
```

Comment:
Should probably move this function to goutput.c and make it behave with the stdout handler for the reasons explained above.